README.md
gocd-vault-secrets
==================

<img height="80" width="80" align="right" src="resources/amperity/gocd/secret/vault/logo.svg"/>

[![CircleCI](https://circleci.com/gh/amperity/gocd-vault-secrets.svg?style=shield&circle-token=b9256c6d46160ab045b44cdfe5add3954dd0cbf2)](https://circleci.com/gh/amperity/gocd-vault-secrets)
[![codecov](https://codecov.io/gh/amperity/gocd-vault-secrets/branch/master/graph/badge.svg)](https://codecov.io/gh/amperity/gocd-vault-secrets)

A plugin for [GoCD](https://www.gocd.org/) providing secret material support via
[HashiCorp Vault](https://www.vaultproject.io/).

This plugin has not yet reached `v1`, but feel free to download a `v0` release (see the installation guide below).


<br/>


## Installation

Releases are published on the [GitHub project](https://github.com/amperity/gocd-vault-secrets/releases).
Download the latest version of the plugin and
[place it in your server's external plugin directory](https://docs.gocd.org/current/extension_points/plugin_user_guide.html).
Once installed, restart the GoCD server to load the plugin. The plugin should
appear in your server's `Admin - Plugins` page when it is back up.


## Configuration

1. Install the Vault Plugin on GoCD as specified in the **Installation** section of this README.
2. Navigate to the Secret Management pane in GoCD.
3. Add a new secret configuration, and fill out the details as desired.
4. Set up the pipeline or environment you wish to use. You can specify secret params as described below.

### Secret Params
In GoCD, you can refer to secret params like this:
```
{{SECRET:[<SECRET CONFIG ID>][<SECRET KEY>]}}
```

If this does not look familiar to you, you may want to check out the [GoCD secret management docs](https://docs.gocd.org/current/configuration/secrets_management.html).

Here's how to set up a secret param using the Vault plugin (please note that these are whitespace and case sensitive):

##### Token Generation
```
{{SECRET:[<ID SPECIFIED IN PART 3>][TOKEN:]}}
```
If you want that token to inherit specific policies, you can specify them after `TOKEN:` and seperated by commas:
```
{{SECRET:[<ID SPECIFIED IN PART 3>][TOKEN:<POLICY1>,<POLICY2>,<POLICY3>]}}
```

##### KV API V1, Generic Secret Engine
```
{{SECRET:[<ID SPECIFIED IN PART 3>][<VAULT PATH>]}}
```
If you wish to access a specific key at that path, you can do so as follows:
```
{{SECRET:[<ID SPECIFIED IN PART 3>][<VAULT PATH>#<KEY>]}}
```
This works with arbitrary levels of nested maps as well:
```
{{SECRET:[<ID SPECIFIED IN PART 3>][<VAULT PATH>#<KEY1>#<KEY2>#<KEY3>]}}
```
where `<VAULT PATH>#<KEY1>` returns a mapping with `KEY2` in it and so on.

##### KV API V2, KV Secret Engine
```
{{SECRET:[<ID SPECIFIED IN PART 3>][<MOUNT>/data/<REST OF VAULT PATH>#data]}}
```
If you wish to access a specific key at that path, you can do so as follows:
```
{{SECRET:[<ID SPECIFIED IN PART 3>][<MOUNT>/data/<REST OF VAULT PATH>#data#<KEY>]}}
```
This works with arbitrary levels of nested maps as well:
```
{{SECRET:[<ID SPECIFIED IN PART 3>][<MOUNT>/data/<REST OF VAULT PATH>#data#<KEY1>#<KEY2>#<KEY3>]}}
```
where `<MOUNT>/data/<REST OF VAULT PATH>#data#<KEY1>` returns a mapping with `KEY2` in it and so on.

## FAQ

##### Why not just store secrets in the pipelines directly?
Operational Complexity caused by Storing Secrets in the Pipelines:
- No coherent form of version management, you should be able to roll back secrets across all pipelines simultaneously and push updates to all pipelines simaultaneously
- Duplicated secrets accross pipelines can cause a lot of problems with consistency. It's difficult to tell whether two secrets across pipelines are the same since they may have different hashes
- Secrets will almost definitely be needed outside of GoCD as well, making multiple different secret sources rely on manually being configured to be the same
- Secrets should be rotated frequently, which is a real pain
- Difficult to tell what secrets are since they can only be decrypted by GoCD

Security Risks caused by Storing Secrets in the Pipelines directly:
- Not safe to store any credentials (even encrypted) in our SCM repos
- Increased attack/leak area since secrets will be duplicated
- Slower response after leaked credentials since it's harder to roll secrets

##### Why not just ust the [GoCD file-based](https://github.com/gocd/gocd-file-based-secrets-plugin) secrets plugin?
Assuming you're using Vault:
- Secrets will almost definitely be needed outside of GoCD as well, making multiple different secret sources rely on manually being configured to be the same
- Increased attack/leak area since secrets will be duplicated
- Slower response after leaked credentials since pipelines it's harder to roll secrets
- The whole point behind Vault is to be a central store for all your secrets, let it do what it does well

If you're not using Vault:
- You may have no other choice but to use the GoCD file-based secrets plugin
- You should consider using some central store to manage your secrets

##### When should I use this plugin or [the other GoCD Vault Secrets Plugin](https://github.com/gocd/gocd-vault-secret-plugin)?
Benifits of this plugin (currently) not implemented in the other plugin:
- Supports AWS IAM Authentication
- Supports logical path secret lookup, thereby supporting every secrets engine
- Supports nested data access within secrets
- Open source so you can add new features easily
- Requires little background knowledge to jump in and add features

Benifits of the other plugin (currently) not implemented in this plugin:
- Supports `approle` and `cert` authentication
- Built specifically for `kvv2` secret engines, so accessing those secrets is simpler

##### I have a new feature idea, what do I do now?
Great! First step is you should open an issue on our repo. You should include whether you are creating the new feature yourself, or would like another dev to build it. Next, feel free to either fork the repo and build it yourself (see the local dev section below) or wait for another contributor to build it.

##### I'm not positive how to use this plugin, who can I ask?
If you don't understand the plugin, that is a bug in our documentation. Please open an issue and we'll update the docs as fast as we can!

## Local Development

If you're planning to work on the plugin code locally, see the [`gocd`](gocd)
directory for running a local GoCD server. The file you will almost definitely be contributing to is: [`src/amperity/gocd/secret/vault/plugin.clj`](./src/amperity/gocd/secret/vault/plugin.clj).

### Setting up a local Vault server

1. First, spin up a dev vault server.
```bash
vault server -dev
```

2. In a new terminal tab/window (so you can keep viewing the Vault logs), set up a `generic` secret store.
```bash
export VAULT_ADDR='http://127.0.0.1:8200'
vault secrets enable --path=<PATH> generic
```

**Important Note**: At this point I will assume all Vault CLI commands are run in a session that has the `VAULT_ADDR` environment variable set to your local Vault instance.

3. Write some testing data to the path from part 2.
```bash
vault write <PATH> <KEY>=<VALUE>
```

4. Ensure you can read that data.
```bash
vault read <PATH>
```

### Configuring GoCD

If you have not completed all the above steps, go back and finish those before configuring GoCD.

1. First, ensure the plugin loaded successfully by checking http://localhost:8153/go/admin/plugins.

2. Then, add a new secret config at http://localhost:8153/go/admin/secret_configs.
    - You will probably be running GoCD locally in a docker container and a Vault docker container. If this is the case, make sure that you set the
    Vault URL configuration option in GoCD to `http://host.docker.internal:<PORT>` where `<PORT>` is the port exposing your running Vault dev server.
    You can find it at the top of your `vault -dev` server logs.
	- You will probably be using the *token* authentication method for local dev.
	If this is the case, copy the root `Vault Token` from your running Vault dev server. You can find it at the top of your dev server logs or by copying the id found from running `vault token lookup`.
	**Note**: Using *token* as your authentication method should not be done in production. While the GoCD plugin does store a hashed version of this token, it is still not recommended for secure systems.


## License

Licensed under the Apache License, Version 2.0. See the [LICENSE](LICENSE) file
for rights and restrictions.


CHANGELOG.md
Change Log
==========

All notable changes to this project will be documented in this file. This change
log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
..

## [0.4.0] - 2020-05-26
- Add ability to configure STS signing region for AWS IAM auth method.
  [#32](https://github.com/amperity/gocd-vault-secrets/pull/32)

## [0.3.3] - 2020-01-09
- Fixed bug where plugin couldn't handle reading secrets from multiple configs.
  [#31](https://github.com/amperity/gocd-vault-secrets/pull/31)

## [0.3.2] - 2019-12-06
- Fixed bug where vault lease timer didn't run so client tokens weren't renewed.
  [#29](https://github.com/amperity/gocd-vault-secrets/pull/29)

## [0.3.1] - 2019-11-22
- Added an option to ignore TTLs specified in Vault and force-read secrets
  [#26](https://github.com/amperity/gocd-vault-secrets/issues/26)

## [0.3.0] - 2019-11-20
- Added Token Lookup option using `TOKENS:<POLICY>,<POLICY>,...`
  [#23](https://github.com/amperity/gocd-vault-secrets/issues/23)

## [0.2.0] - 2019-10-29
- Added AWS Auth method
  [#12](https://github.com/amperity/gocd-vault-secrets/pull/12)
- Fixed bug where secret lookup would fail after server restart
  [#9](https://github.com/amperity/gocd-vault-secrets/issues/9)

## [0.1.0] - 2019-10-22
- Initial plugin release
  [#0](https://media.giphy.com/media/o0eOCNkn7cSD6/giphy.gif)
- Supports one Auth Method, direct Vault tokens
  [#4](https://github.com/amperity/gocd-vault-secrets/pull/4)

[Unreleased]: https://github.com/amperity/gocd-vault-secrets/compare/v0.4.0...HEAD
[0.4.0]: https://github.com/amperity/gocd-vault-secrets/compare/v0.3.3...v0.4.0
[0.3.3]: https://github.com/amperity/gocd-vault-secrets/compare/v0.3.2...v0.3.3
[0.3.2]: https://github.com/amperity/gocd-vault-secrets/compare/v0.3.1...v0.3.2
[0.3.1]: https://github.com/amperity/gocd-vault-secrets/compare/v0.3.0...v0.3.1
[0.3.0]: https://github.com/amperity/gocd-vault-secrets/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/amperity/gocd-vault-secrets/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/amperity/gocd-vault-secrets/releases/tag/v0.1.0


test/amperity/gocd/secret/vault/plugin_api.clj
(ns amperity.gocd.secret.vault.plugin-api
  (:require
    [amperity.gocd.secret.vault.plugin :as plugin]
    [amperity.gocd.secret.vault.util :as u]
    [clojure.test :refer [testing deftest is]]
    [vault.client.mock]
    [vault.core :as vault])
  (:import
    (com.thoughtworks.go.plugin.api.request
      DefaultGoPluginApiRequest)
    (com.thoughtworks.go.plugin.api.response
      DefaultGoPluginApiResponse)))


;; Utils ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(defn default-go-plugin-api-request
  "A .DefaultGoPluginApiRequest with the given name and params.
  Parameters:
  - `request-name` string of the name (essentially the route) of the request
  - `request-params` map of request parameters"
  ([request-name request-params]
   (doto (DefaultGoPluginApiRequest. nil nil request-name)
     (.setRequestParams request-params)))
  ([request-name]
   (default-go-plugin-api-request request-name {})))


(defn response-equal
  "Determines if the given responses are equal enough to be treated the same by the GoCD server.
  Parameters:
  - `response1`: .GoPluginApiResponse to compare to response2
  - `response2`: .GoPluginApiResponse to compare to response1"
  [response1 response2]
  ;; Uncomment these print statements for debugging
  ;(println (u/json-encode response1))
  ;(println (u/json-encode response2))
  (and (= (.responseCode response1) (.responseCode response2))
       (= (.responseBody response1) (.responseBody response2))
       (= (.responseHeaders response1) (.responseHeaders response2))))


(defn mock-client-atom
  "A mock vault client using the secrets found in `resources/secret-fixture.edn`"
  []
  (atom (vault/new-client "mock:amperity/gocd/secret/vault/secret-fixture.edn")))

;; Common Logic Tests ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Handler Tests
(deftest handler-with-nil-response
  (testing "Handler when handle-method returns a GoPluginApiResponse response"
    (with-redefs [plugin/handle-request (fn [_ _ _] {:response-code 200 :response-body "" :response-headers {}})]
      (is (response-equal (DefaultGoPluginApiResponse/success "\"\"")
                          (plugin/handler (mock-client-atom) (default-go-plugin-api-request nil)))))))


(deftest handler-with-plugin-response
  (testing "Handler when handle-method returns a GoPluginApiResponse response"
    (with-redefs [plugin/handle-request
                  (fn [_ _ _] {:response-code 200 :response-body {:message "hello"} :response-headers {}})]
      (is (response-equal (DefaultGoPluginApiResponse/success "{\"message\":\"hello\"}")
                          (plugin/handler (mock-client-atom) (default-go-plugin-api-request nil)))))))


(deftest handler-with-json-response
  (testing "Handler when handle-method returns a json response"
    (let [response {:response-code 200 :response-headers {} :response-body {:try "this"}}]
      (with-redefs [plugin/handle-request (fn [_ _ _] response)]
        (is (response-equal (DefaultGoPluginApiResponse/success "{\"try\":\"this\"}")
                            (plugin/handler (mock-client-atom) (default-go-plugin-api-request nil))))))))

;; Endpoint Tests ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
(deftest get-icon
  (testing "Get icon endpoint with well formed requests"
    (let [result (plugin/handle-request (mock-client-atom) "cd.go.secrets.get-icon" "")
          body (:response-body result)
          status (:response-code result)]
      (is "image/svg+xml"
          (:content_type body))
      (is (:data body))
      (is (= 200 status)))))


(deftest validate
  (with-redefs [vault/authenticate! (fn [_ _ _] true)]
    (testing "Validate correctly handles case with no errors (no false positives)"
      (let [result (plugin/handle-request
                     (mock-client-atom) "go.cd.secrets.validate"
                     {:vault_addr "https://amperity.com"
                      :auth_method "token"
                      :vault_token "abc123"})
            body (:response-body result)
            status (:response-code result)]
        (is (= [] body))
        (is (= 200 status))))
    (testing "Validate correctly handles case with errors (no false negatives, no false positives)"
      (let [result (plugin/handle-request
                     (mock-client-atom) "go.cd.secrets.validate"
                     {:vault_addr "protocol://amperity.com"})
            body (:response-body result)
            status (:response-code result)]
        (is (= [{:key     :vault_addr
                 :message "Vault URL must start with http:// or https://"}
                {:key     :auth_method
                 :message "Authentication Method is required"}]
               body))
        (is (= 200 status))))
    (testing "Validate also resets the vault client if a new URL is specified, when input is valid only"
      (with-redefs [plugin/authenticate-client-from-inputs!
                    (fn [_ i1]
                      (is (= {:auth_method "token"
                              :vault_addr  "https://amperity.com"}
                             i1)))]
        (let [fake-client (atom nil)
              result (plugin/handle-request
                       fake-client "go.cd.secrets.validate"
                       {:vault_addr  "https://amperity.com"
                        :auth_method "token"})
              body (:response-body result)
              status (:response-code result)]
          (is (= 200 status))
          (is (= [] body))
          (is (some? @fake-client)))
        (let [fake-client (atom nil)
              result (plugin/handle-request
                       fake-client "go.cd.secrets.validate"
                       {:vault_addr "https://amperity.com"})
              status (:response-code result)]
          (is (= 200 status))
          (is (nil? @fake-client)))))
    (testing "Validate a does not reset the vault client if no new URL is specified "
      (let [fake-client (atom nil)
            result (plugin/handle-request
                     fake-client "go.cd.secrets.validate"
                     {})
            status (:response-code result)]
        (is (= 200 status))
        (is (nil? @fake-client)))))
  (testing "Validate also returns client authentication errors"
    (let [fake-client (atom nil)
          result (plugin/handle-request
                   fake-client "go.cd.secrets.validate"
                   {:vault_addr  "https://amperity.com"
                    :auth_method "token"})
          body (:response-body result)
          status (:response-code result)]
      (is (= 200 status))
      (is (= [{:key     :auth_method
               :message "Unable to authenticate Vault client:
java.lang.IllegalArgumentException: Token credential must be a string"}]
             body))
      (is (some? @fake-client)))))


(deftest get-metadata
  (testing "Input metadata is returned with the correctly structured response"
    (let [result (plugin/handle-request (mock-client-atom) "go.cd.secrets.get-metadata" {})
          body (:response-body result)
          status (:response-code result)]
      (is (= 200 status))
      (is (not (empty? body)))
      (is (every? :key body))
      (is (every? #(and (contains? % :required) (contains? % :secure)) (map :metadata body))))))


(deftest secrets-lookup
  (testing "Can look up secrets stored in vault given a well formed request"
    (let [result (plugin/handle-request
                   (mock-client-atom)
                   "go.cd.secrets.secrets-lookup"
                   {:configuration {}
                    ;; The keys will likely be string in the http vault client instance,
                    ;; but this is easier for testing.
                    :keys          [:batman :hulk :wonder-woman]})
          body (:response-body result)
          status (:response-code result)]
      (is (= [{:key :batman :value "Bruce Wayne"}
              {:key :hulk :value "Bruce Banner"}
              {:key :wonder-woman :value "Diana Prince"}]
             body))
      (is (= 200 status))))
  (testing "Fails cleanly when looking up secrets that don't exist"
    (let [result (plugin/handle-request (mock-client-atom) "go.cd.secrets.secrets-lookup"
                                        {:configuration {}
                                         :keys          [:dr-who :jack-the-ripper]})
          body (:response-body result)
          status (:response-code result)]
      (is (= {:message "Unable to resolve key(s) [:dr-who :jack-the-ripper]"}
             body))
      (is (= 404 status))))
  (testing "Fails cleanly when other lookup error occurs"
    (let [mock-client-that-errors (reify vault.core/SecretClient
                                    (read-secret [_ _ _] (throw (ex-info "Mock Exception" {}))))
          result (plugin/handle-request
                   (atom mock-client-that-errors)
                   "go.cd.secrets.secrets-lookup"
                   {:configuration {}
                    :keys          [:batman]})
          body (:response-body result)
          status (:response-code result)]
      (is (= {:message "Error occurred during lookup:\nclojure.lang.ExceptionInfo: Mock Exception {}"} body))
      (is (= 500 status)))))


project.clj
(defproject amperity/gocd-vault-secrets "0.4.0"
  :description "A plugin for GoCD providing secret material support via HashiCorp Vault."
  :url "https://github.com/amperity/gocd-vault-secrets"
  :license {:name "Apache License 2.0"
            :url "http://www.apache.org/licenses/LICENSE-2.0"}

  :plugins
  [[lein-cloverage "1.1.0"]]

  :dependencies
  [[amperity/vault-clj "0.7.1"]
   [amperity/vault-clj-aws "0.0.3"]
   [com.google.code.gson/gson "2.8.5"]
   [org.clojure/clojure "1.10.1"]]

  :java-source-paths ["src"]

  :hiera
  {:cluster-depth 4
   :vertical false
   :show-external false}

  :profiles
  {:provided
   {:dependencies
    [[cd.go.plugin/go-plugin-api "19.7.0"]
     [com.google.guava/guava "23.0"]]}

   :repl
   {:source-paths ["dev"]
    :dependencies
    [[org.clojure/tools.namespace "0.2.11"]]}

   :uberjar
   {:target-path "target/uberjar"
    :aot :all}})


src/amperity/gocd/secret/vault/util.clj
(ns amperity.gocd.secret.vault.util
  "Plugin utilities."
  (:require
    [clojure.java.io :as io]
    [clojure.walk :as walk]
    [clojure.xml :as xml])
  (:import
    com.google.gson.Gson
    java.util.Base64))


(def plugin-id
  "Unique plugin identifier string."
  (-> (io/resource "plugin.xml")
      (io/input-stream)
      (xml/parse)
      (get-in [:attrs :id])))


(def ^:private ^Gson gson
  (Gson.))


(defn- stringify-keywords
  "Recursively transforms all keywords into strings."
  [v]
  (walk/postwalk
    (fn [x]
      (if (keyword? x)
        (subs (str x) 1)
        x))
    v))


(defn- gson->clj
  "Recursively transforms data structures into their Clojure equivalents."
  [v]
  (walk/prewalk
    (fn [x]
      (cond
        (instance? java.util.List x)
        (into [] x)

        (instance? java.util.Map x)
        (into {}
              (map (fn [[k v]]
                     (if (string? k)
                       [(keyword k) v]
                       [k v])))
              x)

        (instance? java.util.Set x)
        (into #{} x)

        :else x))
    v))


(defn json-encode
  "Encode a value to a JSON string."
  ^String
  [value]
  (.toJson gson (stringify-keywords value)))


(defn json-decode-map
  "Decode a map from a JSON string."
  [^String json]
  (gson->clj (.fromJson gson json java.util.Map)))


(defn b64-encode-str
  "Encode the bytes in a string to a Base64-encoded string."
  [^String data]
  (.encodeToString (Base64/getEncoder) (.getBytes data)))


src/amperity/gocd/secret/vault/plugin.clj
(ns amperity.gocd.secret.vault.plugin
  "Core plugin implementation."
  (:require
    [amperity.gocd.secret.vault.logging :as log]
    [amperity.gocd.secret.vault.util :as u]
    [clojure.java.io :as io]
    [clojure.string :as str]
    [vault.client.ext.aws]
    [vault.client.http]
    [vault.core :as vault])
  (:import
    clojure.lang.ExceptionInfo
    (com.amazonaws.auth
      AWSCredentials)
    (com.thoughtworks.go.plugin.api.exceptions
      UnhandledRequestTypeException)
    (com.thoughtworks.go.plugin.api.request
      GoPluginApiRequest)
    (com.thoughtworks.go.plugin.api.response
      DefaultGoPluginApiResponse)))


;; ## Plugin Initialization

(defn initialize!
  "Set up the vault client."
  [logger app-accessor]
  (alter-var-root #'log/logger (constantly logger)))


;; ## Model

;; A map of user configurable fields to the all the data necessary to define those fields
(def input-schema
  {;; the name of the corresponding input field
   :vault_addr  {;; the metadata required by the plugin API about every input field
                 :metadata     {:required true :secure false}
                 ;; the string representation of the URL
                 :label        "Vault URL"
                 ;; an array outlining extended validation the function may require
                 :validate-fns [;; nil if no error detected, else a string descring the error
                                #(when-not (or (str/starts-with? % "http://") (str/starts-with? % "https://"))
                                   "Vault URL must start with http:// or https://")]}
   :auth_method {:metadata     {:required true :secure false}
                 :label        "Authentication Method"
                 :validate-fns []}
   :force_read  {:metadata     {:required true :secure false}
                 :label         "Force Read (Secret Caching)"
                 :validate-fns []}
   ;; Token Auth
   :vault_token {:metadata     {:required false :secure true}
                 :label        "Vault Token"
                 :validate-fns [#(when-not (string? %)
                                   "Vault Token must be a string")]}
   ;; AWS Auth
   :iam_role {:metadata {:required false :secure false}
              :label "IAM Role"
              :validate-fns []}

   :sts_region {:metadata {:required false :secure false}
                :label "STS Signing Region"
                :validate-fns []}})

;; Signifies a token creation instead of a secret lookup
(def signify-token-creation-str "TOKEN:")

;; ## Request Handling

(defmulti handle-request
  "Handle a plugin API request and respond. Methods should return a map containing the following 3 keys:
  ```
  {:response-code     <int: the returned status, follows HTTP status conventions>
   :response-body     <json-coercible: the response body, will be coerced into JSON>
   :response-headers  <map: the response headers, follows HTTP header conventions>}
  ```

  Params:
  - `client`: Atom(vault.client), used for auth and retrieval of all the secret values
  - `req-name`: string, determines how to dispatch among implementing methods, essentially the route
  - `data`: map, the body of the message passed from the GoCD server"
  (fn dispatch
    [req-name data]
    ;(log/logger "info" (str "Vault plugin received: " req-name) nil)  ;; Most useful log statement for debugging
    req-name))


(defmethod handle-request :default
  [req-name _]
  (throw (UnhandledRequestTypeException. req-name)))


(defn handler
  "Request handling entry-point."
  [^GoPluginApiRequest request]
  (try
    (let [req-name (.requestName request)
          req-data (when-not (str/blank? (.requestBody request))
                     (u/json-decode-map (.requestBody request)))
          {status :response-code body :response-body headers :response-headers} (handle-request req-name req-data)]
      (DefaultGoPluginApiResponse. status (u/json-encode body) headers))
    (catch UnhandledRequestTypeException ex
      (throw ex))
    (catch Exception ex
      (log/errorx ex "Failed to process %s plugin request%s"
                  (.requestName request)
                  (when-let [data (not-empty (ex-data ex))]
                    (str " " (pr-str data))))
      (DefaultGoPluginApiResponse/error (.getMessage ex)))))


;; ## Plugin Metadata

;; This call is expected to return the icon for the plugin, so as to make
;; it easy for users to identify the plugin.
(defmethod handle-request "go.cd.secrets.get-icon"
  [_ _]
  (let [icon-svg (slurp (io/resource "amperity/gocd/secret/vault/logo.svg"))]
    {:response-code    200
     :response-headers {}
     :response-body    {:content_type "image/svg+xml"
                        :data         (u/b64-encode-str icon-svg)}}))


;; ## Secrets Configuration

;; This message should return an HTML template allowing users to configure the
;; secret backend in GoCD.
(defmethod handle-request "go.cd.secrets.secrets-config.get-view"
  [_ _]
  (let [view-html (slurp (io/resource "amperity/gocd/secret/vault/secrets-view.html"))]
    {:response-code    200
     :response-headers {}
     :response-body    {:template view-html}}))


;; This message should return metadata about the available settings for
;; configuring a secret backend in GoCD.
(defmethod handle-request "go.cd.secrets.secrets-config.get-metadata"
  [_ _]
  {:response-code    200
   :response-headers {}
   :response-body    (mapv (fn [input-key]
                             {:key      input-key
                              :metadata (-> input-schema input-key :metadata)})
                           (keys input-schema))})


(defn- input-error-message
  "If the input field is valid, does nothing, if it's not, returns a string describing the error.

  Params:
  - `field-key`: the input field you are validating
  - `field-value`: the inputted value which you wish to verify"
  [field-key field-value]
  (let [field-model (field-key input-schema)]
    (if (and (-> field-model :metadata :required) (str/blank? field-value))
      ;; field is required but empty
      (str (:label field-model) " is required")
      (when-not (str/blank? field-value)
        (some #(% field-value) (:validate-fns field-model))))))


(defn- authenticate-client-from-inputs!
  "Returns an authenticated Vault Client from plugin settings.

  Params:
  - `inputs` A map containing the user inputted settings for the plugin"
  [inputs]
  (let [client (vault/new-client (:vault_addr inputs))]
    (case (:auth_method inputs)
      "token"
      (vault/authenticate! client :token
                           (:vault_token inputs))

      "aws-iam"
      (vault/authenticate! client :aws-iam
                           {:iam-role    (:iam_role inputs)
                            :sts-region  (:sts_region inputs)
                            :credentials ^AWSCredentials (:aws_credentials inputs)})

      (throw (ex-info "Unhandled vault auth type"
                      {:user-input (:auth_method inputs)})))

    client))


;; This call is expected to validate the user inputs that form a part of
;; the secret backend configuration.
(defmethod handle-request "go.cd.secrets.secrets-config.validate"
  [_ data]
  (let [input-error (fn [field-key]
                      (when-let [error-message (input-error-message field-key (field-key data))]
                        {:key field-key
                         :message error-message}))
        errors-found (keep input-error (keys input-schema))]
    (if (empty? errors-found)
      (try
        (authenticate-client-from-inputs! data)
        {:response-code 200
         :response-headers {}
         :response-body []}
        (catch Exception ex
          {:response-code 200
           :response-headers {}
           :response-body [{:key :auth_method
                            :message (str "Unable to authenticate Vault client:\n" ex)}]}))
      {:response-code    200
       :response-headers {}
       :response-body    (into [] errors-found)})))


;; ## Secret Usage
(defn- lookup-secrets
  "Creates an lazy seq containing maps specifying GoCD lookup keys and their associated value. Structured:
   ({:key <GoCD lookup key>
     :value <Associated value>}
     ...)

  Params:
  - `client`: The vault.client (*not* as an Atom) you want to use to access Vault
  - `gocd-lookup-keys`: A seq of strings, (<PATH>#<KEY> ...), where <PATH> corresponds to a Vault Path, and <KEY>
  a key found at that path.
  - `force-read`: Ignore secret TTLs and read from Vault regardless of cached secrets?"
  [client gocd-lookup-keys force-read]
  (let [;; A map from Vault paths (as keywords) to their corresponding secret data (defaults to nil)
        paths-to-vals
        (->> gocd-lookup-keys
             (map #(first (str/split % #"#")))
             (into #{})
             (map (fn [path]
                    [(keyword path) (vault/read-secret client path {:not-found nil
                                                                    :force-read force-read})]))
             (into {}))]
    (map
      (fn [gocd-lookup-key]
        {:key   gocd-lookup-key
         :value (get-in paths-to-vals (map keyword (str/split gocd-lookup-key #"#")))})
      gocd-lookup-keys)))


(defn- create-tokens!
  "Creates an lazy seq containing maps specifying GoCD token-keys and their associated tokens. Structured:
 ({:key <GoCD lookup key>
   :value <Associated value>}
   ...)

  Params:
  - `client`: The vault.client (*not* as an Atom) you want to use to access Vault
  - `gocd-lookup-keys`: A seq of strings, (POLICIES:<POLICY>,<POLICY>,<POLICY> ...), where <POLICY> is a policy you
  wish the generated token to have"
  [client token-keys]
  (map
    (fn [token-key]
      {:key token-key
       :value (:client-token (vault/create-token!
                               client
                               {:policies (str/split (subs token-key (count signify-token-creation-str)) #",")}))})
    token-keys))

;; This message is a request to the plugin to look up for secrets for a given
;; list of keys. In addition to the list of keys in the JSON request, the
;; request body will also have the configuration required to connect and lookup
;; for secrets from the external Secret Manager.
(defmethod handle-request "go.cd.secrets.secrets-lookup"
  [_ data]
  (try
    (let [client (authenticate-client-from-inputs! (:configuration data))
          {token-keys true secrets-keys false} (group-by #(str/starts-with?  % signify-token-creation-str) (:keys data))
          secrets (lookup-secrets client secrets-keys (= (-> data :configuration :force_read) "true"))]
      (if-let [missing-keys (->> secrets
                                 (remove :value)
                                 (mapv :key)
                                 not-empty)]
        {:response-code    404
         :response-headers {}
         :response-body    {:message (str "Unable to resolve key(s) " missing-keys)}}
        {:response-code    200
         :response-headers {}
         :response-body    (-> (create-tokens! client token-keys)
                               (concat secrets)
                               (->> (mapv #(update % :value str))))}))
    (catch ExceptionInfo ex
      {:response-code    500
       :response-headers {}
       :response-body    {:message (str "Error occurred during lookup of: " (:keys data) "\n" ex)}})))
